pkg/streamingpromql/operators/aggregations/count_values.go

```diff
@@ -0,0 +1,247 @@
+	"sync"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+type CountValues struct {
+	Inner                    types.InstantVectorOperator
+	LabelName                types.StringOperator
+	TimeRange                types.QueryTimeRange
+	Grouping                 []string // If this is a 'without' aggregation, NewAggregation will ensure that this slice contains __name__.
+	Without                  bool
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+
+	expressionPosition posrange.PositionRange
+
+	resolvedLabelName string
+
+	series [][]promql.FPoint
+
+	// Reuse instances used to generate series labels rather than recreating them every time.
+	labelsBuilder     *labels.Builder
+	labelsBytesBuffer []byte
+	valueBuffer       []byte
+}
+
+var _ types.InstantVectorOperator = &CountValues{}
+
+func NewCountValues(
+	inner types.InstantVectorOperator,
+	labelName types.StringOperator,
+	timeRange types.QueryTimeRange,
+	grouping []string,
+	without bool,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	expressionPosition posrange.PositionRange,
+) *CountValues {
+	return &CountValues{
+		Inner:                    inner,
+		LabelName:                labelName,
+		TimeRange:                timeRange,
+		Grouping:                 grouping,
+		Without:                  without,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+		expressionPosition:       expressionPosition,
+	}
+}
+
+type countValuesSeries struct {
+	labels           labels.Labels
+	outputPointCount int
+	count            []int // One entry per timestamp.
+}
+
+var countValuesSeriesPool = sync.Pool{
+	New: func() interface{} {
+		return &countValuesSeries{}
+	},
+}
+
+func (c *CountValues) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	if err := c.loadLabelName(); err != nil {
+		return nil, err
+	}
+
+	innerMetadata, err := c.Inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer types.PutSeriesMetadataSlice(innerMetadata)
+
+	c.labelsBuilder = labels.NewBuilder(labels.EmptyLabels())
+	c.labelsBytesBuffer = make([]byte, 0, 1024) // Why 1024 bytes? It's what labels.Labels.String() uses as a buffer size, so we use that as a sensible starting point too.
+	defer func() {
+		// Don't hold onto the instances used to generate series labels for longer than necessary.
+		c.labelsBuilder = nil
+		c.labelsBytesBuffer = nil
+		c.valueBuffer = nil
+	}()
+
+	accumulator := map[string]*countValuesSeries{}
+
+	for _, s := range innerMetadata {
+		data, err := c.Inner.NextSeries(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, p := range data.Floats {
+			if err := c.incrementCount(s.Labels, p.T, c.formatFloatValue(p.F), accumulator); err != nil {
```

Comment:
Would caching `c.formatFloatValue(p.F)` going to be beneficial? I will imagine if we have many same values across the result, this will be formatted multiple times for that same values.

Reply:
The benchmarks show this implementation is already significantly faster than Prometheus' engine for the case where there are many series with the same value, and caching like this would make the worst case (many series with different values) slower, so I'm tempted to leave things as-is.